In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("car_data_v3.csv")

In [3]:
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

                     torque  seats  
0            190Nm@ 2000rpm    5.0  
1       250Nm@ 1500-2500rpm    5.0  
2     12.7@ 2,700(kgm@ rpm)    5.0  
3  22.4 kgm at 1750-2750rpm    5.0  
4     11.5@ 4,500(kgm@ rpm)    5.0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB


In [5]:
df.describe()

year  selling_price     km_driven        seats
count  8128.000000   8.128000e+03  8.128000e+03  7907.000000
mean   2013.804011   6.382718e+05  6.981951e+04     5.416719
std       4.044249   8.062534e+05  5.655055e+04     0.959588
min    1983.000000   2.999900e+04  1.000000e+00     2.000000
25%    2011.000000   2.549990e+05  3.500000e+04     5.000000
50%    2015.000000   4.500000e+05  6.000000e+04     5.000000
75%    2017.000000   6.750000e+05  9.800000e+04     5.000000
max    2020.000000   1.000000e+07  2.360457e+06    14.000000

In [6]:
df.isnull().sum()

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        215
torque           222
seats            221
dtype: int64

In [7]:
df.columns

Index(['name', 'year', 'selling_price', 'km_driven', 'fuel', 'seller_type',
       'transmission', 'owner', 'mileage', 'engine', 'max_power', 'torque',
       'seats'],
      dtype='object')

In [8]:
df['seats'] = df['seats'].fillna(df['seats'].median())

In [9]:
missing_cols = ['mileage', 'engine', 'max_power', 'torque']
for i in missing_cols:
    df[i] = df[i].fillna(df[i].mode()[0])

In [10]:
df.isna().sum()

name             0
year             0
selling_price    0
km_driven        0
fuel             0
seller_type      0
transmission     0
owner            0
mileage          0
engine           0
max_power        0
torque           0
seats            0
dtype: int64

In [11]:
df1 = df.copy()

In [12]:
df1.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

                     torque  seats  
0            190Nm@ 2000rpm    5.0  
1       250Nm@ 1500-2500rpm    5.0  
2     12.7@ 2,700(kgm@ rpm)    5.0  
3  22.4 kgm at 1750-2750rpm    5.0  
4     11.5@ 4,500(kgm@ rpm)    5.0

In [13]:
df['seats'].unique()

array([ 5.,  4.,  7.,  8.,  6.,  9., 10., 14.,  2.])

In [14]:
df1.drop(["name"], axis = 1, inplace = True)

In [15]:
df1.head()

year  selling_price  km_driven    fuel seller_type transmission  \
0  2014         450000     145500  Diesel  Individual       Manual   
1  2014         370000     120000  Diesel  Individual       Manual   
2  2006         158000     140000  Petrol  Individual       Manual   
3  2010         225000     127000  Diesel  Individual       Manual   
4  2007         130000     120000  Petrol  Individual       Manual   

          owner     mileage   engine   max_power                    torque  \
0   First Owner   23.4 kmpl  1248 CC      74 bhp            190Nm@ 2000rpm   
1  Second Owner  21.14 kmpl  1498 CC  103.52 bhp       250Nm@ 1500-2500rpm   
2   Third Owner   17.7 kmpl  1497 CC      78 bhp     12.7@ 2,700(kgm@ rpm)   
3   First Owner   23.0 kmpl  1396 CC      90 bhp  22.4 kgm at 1750-2750rpm   
4   First Owner   16.1 kmpl  1298 CC    88.2 bhp     11.5@ 4,500(kgm@ rpm)   

   seats  
0    5.0  
1    5.0  
2    5.0  
3    5.0  
4    5.0

In [16]:
df1[df1['mileage'].str.contains("km/kg")]

year  selling_price  km_driven fuel seller_type transmission  \
6     2007          96000     175000  LPG  Individual       Manual   
35    2019         330000      10000  CNG  Individual       Manual   
90    2010         225000      44000  LPG      Dealer       Manual   
203   2016         360000      50000  CNG  Individual       Manual   
402   2020         409999      35000  CNG  Individual       Manual   
...    ...            ...        ...  ...         ...          ...   
7308  2010         155000      70000  CNG  Individual       Manual   
7543  2019         330000      10000  CNG  Individual       Manual   
7642  2012         300000      70000  LPG  Individual       Manual   
7733  2015         380000      20000  CNG  Individual       Manual   
7913  2011         209000          1  CNG  Individual       Manual   

                     owner      mileage   engine  max_power  \
6              First Owner   17.3 km/kg  1061 CC   57.5 bhp   
35            Second Owner  33.44 km/kg   796 CC   40.3 bhp   
90             First Owner   26.2 km/kg   998 CC   58.2 bhp   
203            First Owner   26.6 km/kg   998 CC  58.16 bhp   
402            First Owner  20.88 km/kg  1196 CC   61.7 bhp   
...                    ...          ...      ...        ...   
7308          Second Owner  26.83 km/kg   796 CC   38.4 bhp   
7543          Second Owner  33.44 km/kg   796 CC   40.3 bhp   
7642           First Owner   13.2 km/kg   995 CC   60.2 bhp   
7733           First Owner  21.94 km/kg  1196 CC   61.7 bhp   
7913  Fourth & Above Owner   15.1 km/kg  1196 CC     73 bhp   

                    torque  seats  
6     7.8@ 4,500(kgm@ rpm)    5.0  
35           60Nm@ 3500rpm    4.0  
90           77Nm@ 3500rpm    5.0  
203          77Nm@ 3500rpm    5.0  
402          85Nm@ 3000rpm    5.0  
...                    ...    ...  
7308         62Nm@ 3000rpm    5.0  
7543         60Nm@ 3500rpm    4.0  
7642       90.3Nm@ 4200rpm    5.0  
7733         85Nm@ 3000rpm    5.0  
7913        101Nm@ 3000rpm    5.0  

[88 rows x 12 columns]

In [17]:
df1['mileage'].loc[0].split(" ")[0]

'23.4'

In [18]:
df1['Mileage'] = df1['mileage'].str.split(" ").str[0]

In [19]:
df1['Mileage'] = df1['Mileage'].astype("float64")

In [20]:
df1.head(2)

year  selling_price  km_driven    fuel seller_type transmission  \
0  2014         450000     145500  Diesel  Individual       Manual   
1  2014         370000     120000  Diesel  Individual       Manual   

          owner     mileage   engine   max_power               torque  seats  \
0   First Owner   23.4 kmpl  1248 CC      74 bhp       190Nm@ 2000rpm    5.0   
1  Second Owner  21.14 kmpl  1498 CC  103.52 bhp  250Nm@ 1500-2500rpm    5.0   

   Mileage  
0    23.40  
1    21.14

In [21]:
df1['engine'].unique()

array(['1248 CC', '1498 CC', '1497 CC', '1396 CC', '1298 CC', '1197 CC',
       '1061 CC', '796 CC', '1364 CC', '1399 CC', '1461 CC', '993 CC',
       '1198 CC', '1199 CC', '998 CC', '1591 CC', '2179 CC', '1368 CC',
       '2982 CC', '2494 CC', '2143 CC', '2477 CC', '1462 CC', '2755 CC',
       '1968 CC', '1798 CC', '1196 CC', '1373 CC', '1598 CC', '1998 CC',
       '1086 CC', '1194 CC', '1172 CC', '1405 CC', '1582 CC', '999 CC',
       '2487 CC', '1999 CC', '3604 CC', '2987 CC', '1995 CC', '1451 CC',
       '1969 CC', '2967 CC', '2497 CC', '1797 CC', '1991 CC', '2362 CC',
       '1493 CC', '1599 CC', '1341 CC', '1794 CC', '799 CC', '1193 CC',
       '2696 CC', '1495 CC', '1186 CC', '1047 CC', '2498 CC', '2956 CC',
       '2523 CC', '1120 CC', '624 CC', '1496 CC', '1984 CC', '2354 CC',
       '814 CC', '793 CC', '1799 CC', '936 CC', '1956 CC', '1997 CC',
       '1499 CC', '1948 CC', '2997 CC', '2489 CC', '2499 CC', '2609 CC',
       '2953 CC', '1150 CC', '1994 CC', '1388 CC', '1527 CC'

In [22]:
df1['Engine_CC'] = df1['engine'].str.split(" ").str[0].astype("int64")

In [23]:
df1.head(2)

year  selling_price  km_driven    fuel seller_type transmission  \
0  2014         450000     145500  Diesel  Individual       Manual   
1  2014         370000     120000  Diesel  Individual       Manual   

          owner     mileage   engine   max_power               torque  seats  \
0   First Owner   23.4 kmpl  1248 CC      74 bhp       190Nm@ 2000rpm    5.0   
1  Second Owner  21.14 kmpl  1498 CC  103.52 bhp  250Nm@ 1500-2500rpm    5.0   

   Mileage  Engine_CC  
0    23.40       1248  
1    21.14       1498

In [24]:
df[~df1['max_power'].str.contains("bhp")].index

Index([575, 576, 1442, 1443, 2549, 2550], dtype='int64')

In [25]:
duplicate_indices = [576, 1443, 2550]
df1 = df1.drop(index=duplicate_indices).reset_index(drop=True)

In [26]:
df1[df1['max_power'] == "0"]

year  selling_price  km_driven    fuel seller_type transmission  \
575   2011         204999      97500  Petrol  Individual       Manual   
1441  2017         589000      41232  Diesel      Dealer       Manual   
2547  2012         240000      70000  Diesel  Individual       Manual   

            owner    mileage   engine max_power          torque  seats  \
575   First Owner  18.9 kmpl  1248 CC         0  190Nm@ 2000rpm    5.0   
1441  First Owner  18.9 kmpl  1248 CC         0  190Nm@ 2000rpm    5.0   
2547  First Owner  18.9 kmpl  1248 CC         0  190Nm@ 2000rpm    5.0   

      Mileage  Engine_CC  
575      18.9       1248  
1441     18.9       1248  
2547     18.9       1248

In [27]:
df1['max_power'] = df1['max_power'].replace("0", np.nan)

In [28]:
df1.isna().sum()

year             0
selling_price    0
km_driven        0
fuel             0
seller_type      0
transmission     0
owner            0
mileage          0
engine           0
max_power        3
torque           0
seats            0
Mileage          0
Engine_CC        0
dtype: int64

In [29]:
df1['max_power'] = df1['max_power'].fillna(df1['max_power'].mode()[0])
df1.isna().sum()

year             0
selling_price    0
km_driven        0
fuel             0
seller_type      0
transmission     0
owner            0
mileage          0
engine           0
max_power        0
torque           0
seats            0
Mileage          0
Engine_CC        0
dtype: int64

In [30]:
df1['Max_power'] = df1['max_power'].str.split(" ").str[0]

In [48]:
df1[df1["Max_power"] == ""]

Empty DataFrame
Columns: [year, selling_price, km_driven, fuel, seller_type, transmission, owner, mileage, engine, max_power, torque, seats, Mileage, Engine_CC, Max_power]
Index: []

In [46]:
df1.drop(index = [4930], inplace = True)

In [47]:
df1[df1["Max_power"] == ""]
df1['Max_power'] = df1['Max_power'].astype("float64")

In [49]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8124 entries, 0 to 8124
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           8124 non-null   int64  
 1   selling_price  8124 non-null   int64  
 2   km_driven      8124 non-null   int64  
 3   fuel           8124 non-null   object 
 4   seller_type    8124 non-null   object 
 5   transmission   8124 non-null   object 
 6   owner          8124 non-null   object 
 7   mileage        8124 non-null   object 
 8   engine         8124 non-null   object 
 9   max_power      8124 non-null   object 
 10  torque         8124 non-null   object 
 11  seats          8124 non-null   float64
 12  Mileage        8124 non-null   float64
 13  Engine_CC      8124 non-null   int64  
 14  Max_power      8124 non-null   float64
dtypes: float64(3), int64(4), object(8)
memory usage: 1015.5+ KB


In [50]:
df3 = df1.copy()

In [51]:
df3['Torque'] = df3['torque'].str.split(" ").str[0]

In [52]:
len(df3['Torque'])

8124

In [53]:

import re

# Apply regex to keep only digits in each string value
df3['Torque'] = df3['Torque'].astype(str).apply(lambda x: re.sub(r'\D', '', x))


In [54]:
len(df3['Torque'])

8124

In [55]:
df3['Torque'] = df3['Torque'].astype("float64")

In [56]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8124 entries, 0 to 8124
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           8124 non-null   int64  
 1   selling_price  8124 non-null   int64  
 2   km_driven      8124 non-null   int64  
 3   fuel           8124 non-null   object 
 4   seller_type    8124 non-null   object 
 5   transmission   8124 non-null   object 
 6   owner          8124 non-null   object 
 7   mileage        8124 non-null   object 
 8   engine         8124 non-null   object 
 9   max_power      8124 non-null   object 
 10  torque         8124 non-null   object 
 11  seats          8124 non-null   float64
 12  Mileage        8124 non-null   float64
 13  Engine_CC      8124 non-null   int64  
 14  Max_power      8124 non-null   float64
 15  Torque         8124 non-null   float64
dtypes: float64(4), int64(4), object(8)
memory usage: 1.1+ MB


In [57]:
df3.head()

year  selling_price  km_driven    fuel seller_type transmission  \
0  2014         450000     145500  Diesel  Individual       Manual   
1  2014         370000     120000  Diesel  Individual       Manual   
2  2006         158000     140000  Petrol  Individual       Manual   
3  2010         225000     127000  Diesel  Individual       Manual   
4  2007         130000     120000  Petrol  Individual       Manual   

          owner     mileage   engine   max_power                    torque  \
0   First Owner   23.4 kmpl  1248 CC      74 bhp            190Nm@ 2000rpm   
1  Second Owner  21.14 kmpl  1498 CC  103.52 bhp       250Nm@ 1500-2500rpm   
2   Third Owner   17.7 kmpl  1497 CC      78 bhp     12.7@ 2,700(kgm@ rpm)   
3   First Owner   23.0 kmpl  1396 CC      90 bhp  22.4 kgm at 1750-2750rpm   
4   First Owner   16.1 kmpl  1298 CC    88.2 bhp     11.5@ 4,500(kgm@ rpm)   

   seats  Mileage  Engine_CC  Max_power  Torque  
0    5.0    23.40       1248      74.00   190.0  
1    5.0    21.14       1498     103.52   250.0  
2    5.0    17.70       1497      78.00   127.0  
3    5.0    23.00       1396      90.00   224.0  
4    5.0    16.10       1298      88.20   115.0

In [58]:
df3['torque'].str.split(" ").str[1]

0                2000rpm
1           1500-2500rpm
2             2,700(kgm@
3                    kgm
4             4,500(kgm@
              ...       
8120             4000rpm
8121    1,900-2,750(kgm@
8122             2000rpm
8123        1800-3000rpm
8124        1800-3000rpm
Name: torque, Length: 8124, dtype: object

In [60]:
df3['RPM'] = df3['torque'].str.split(" ").str[1]

In [61]:
df3['RPM']

0                2000rpm
1           1500-2500rpm
2             2,700(kgm@
3                    kgm
4             4,500(kgm@
              ...       
8120             4000rpm
8121    1,900-2,750(kgm@
8122             2000rpm
8123        1800-3000rpm
8124        1800-3000rpm
Name: RPM, Length: 8124, dtype: object

In [62]:
import re

def extract_valid_format(x):
    x = str(x).strip().lower()
    x = x.replace(',', '')  # Remove commas

    # Match and extract '2000-3000rpm' → keep '2000-3000'
    match_range = re.search(r'(\d+\s*-\s*\d+)\s*rpm', x)
    if match_range:
        return match_range.group(1)

    # Match and extract '2000rpm' → keep '2000'
    match_single = re.search(r'(\d+)\s*rpm', x)
    if match_single:
        return match_single.group(1)

    # Fallback: just extract a plain number like 2700 from '2,700(kgm@'
    match_plain = re.search(r'\d+', x)
    if match_plain:
        return match_plain.group(0)

    return ""

df3['RPM'] = df3['RPM'].apply(extract_valid_format)

In [63]:
df3['RPM']

0            2000
1       1500-2500
2            2700
3                
4            4500
          ...    
8120         4000
8121         1900
8122         2000
8123    1800-3000
8124    1800-3000
Name: RPM, Length: 8124, dtype: object

In [64]:
len(df3['RPM'].loc[0].split("-"))

1

In [65]:
def range_to_num_1(x):
    tokens = x.split("-")
    if len(tokens) == 2:
        return str((float(tokens[0]) + float(tokens[1])) / 2)
    else:
        return x

# Apply the function to the 'RPM' column
df3['RPM'] = df3['RPM'].apply(lambda x: range_to_num_1(x))

In [66]:
df3['RPM']

0         2000
1       2000.0
2         2700
3             
4         4500
         ...  
8120      4000
8121      1900
8122      2000
8123    2400.0
8124    2400.0
Name: RPM, Length: 8124, dtype: object

In [67]:
df3['RPM'].isna().sum()

np.int64(0)

In [68]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8124 entries, 0 to 8124
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           8124 non-null   int64  
 1   selling_price  8124 non-null   int64  
 2   km_driven      8124 non-null   int64  
 3   fuel           8124 non-null   object 
 4   seller_type    8124 non-null   object 
 5   transmission   8124 non-null   object 
 6   owner          8124 non-null   object 
 7   mileage        8124 non-null   object 
 8   engine         8124 non-null   object 
 9   max_power      8124 non-null   object 
 10  torque         8124 non-null   object 
 11  seats          8124 non-null   float64
 12  Mileage        8124 non-null   float64
 13  Engine_CC      8124 non-null   int64  
 14  Max_power      8124 non-null   float64
 15  Torque         8124 non-null   float64
 16  RPM            8124 non-null   object 
dtypes: float64(4), int64(4), object(9)
memory usage: 1.4+ MB


In [69]:
df3['RPM'] = df3['RPM'].replace("", np.nan)

In [70]:
df3['RPM'] = df3['RPM'].fillna(df3['RPM'].mode()[0])

In [71]:
df3['RPM'] = df3['RPM'].astype("float64")

In [72]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8124 entries, 0 to 8124
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           8124 non-null   int64  
 1   selling_price  8124 non-null   int64  
 2   km_driven      8124 non-null   int64  
 3   fuel           8124 non-null   object 
 4   seller_type    8124 non-null   object 
 5   transmission   8124 non-null   object 
 6   owner          8124 non-null   object 
 7   mileage        8124 non-null   object 
 8   engine         8124 non-null   object 
 9   max_power      8124 non-null   object 
 10  torque         8124 non-null   object 
 11  seats          8124 non-null   float64
 12  Mileage        8124 non-null   float64
 13  Engine_CC      8124 non-null   int64  
 14  Max_power      8124 non-null   float64
 15  Torque         8124 non-null   float64
 16  RPM            8124 non-null   float64
dtypes: float64(5), int64(4), object(8)
memory usage: 1.4+ MB


In [73]:
df.head()

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2      Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4        Maruti Swift VXI BSIII  2007         130000     120000  Petrol   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   Third Owner   17.7 kmpl  1497 CC      78 bhp   
3  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   
4  Individual       Manual   First Owner   16.1 kmpl  1298 CC    88.2 bhp   

                     torque  seats  
0            190Nm@ 2000rpm    5.0  
1       250Nm@ 1500-2500rpm    5.0  
2     12.7@ 2,700(kgm@ rpm)    5.0  
3  22.4 kgm at 1750-2750rpm    5.0  
4     11.5@ 4,500(kgm@ rpm)    5.0

In [74]:
df3.columns

Index(['year', 'selling_price', 'km_driven', 'fuel', 'seller_type',
       'transmission', 'owner', 'mileage', 'engine', 'max_power', 'torque',
       'seats', 'Mileage', 'Engine_CC', 'Max_power', 'Torque', 'RPM'],
      dtype='object')

In [75]:
df3.drop(columns = ['mileage', 'engine', 'max_power', 'torque'],axis = 1, inplace = True)

In [76]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8124 entries, 0 to 8124
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           8124 non-null   int64  
 1   selling_price  8124 non-null   int64  
 2   km_driven      8124 non-null   int64  
 3   fuel           8124 non-null   object 
 4   seller_type    8124 non-null   object 
 5   transmission   8124 non-null   object 
 6   owner          8124 non-null   object 
 7   seats          8124 non-null   float64
 8   Mileage        8124 non-null   float64
 9   Engine_CC      8124 non-null   int64  
 10  Max_power      8124 non-null   float64
 11  Torque         8124 non-null   float64
 12  RPM            8124 non-null   float64
dtypes: float64(5), int64(4), object(4)
memory usage: 1.1+ MB


In [77]:
df3.head()

year  selling_price  km_driven    fuel seller_type transmission  \
0  2014         450000     145500  Diesel  Individual       Manual   
1  2014         370000     120000  Diesel  Individual       Manual   
2  2006         158000     140000  Petrol  Individual       Manual   
3  2010         225000     127000  Diesel  Individual       Manual   
4  2007         130000     120000  Petrol  Individual       Manual   

          owner  seats  Mileage  Engine_CC  Max_power  Torque     RPM  
0   First Owner    5.0    23.40       1248      74.00   190.0  2000.0  
1  Second Owner    5.0    21.14       1498     103.52   250.0  2000.0  
2   Third Owner    5.0    17.70       1497      78.00   127.0  2700.0  
3   First Owner    5.0    23.00       1396      90.00   224.0  2000.0  
4   First Owner    5.0    16.10       1298      88.20   115.0  4500.0

In [78]:
df3['fuel'].unique()

array(['Diesel', 'Petrol', 'LPG', 'CNG'], dtype=object)

In [81]:
cat_cols = df3.select_dtypes(include="object").columns
cat_cols

Index(['fuel', 'seller_type', 'transmission', 'owner'], dtype='object')

In [83]:
for col in cat_cols:
    print(df3[col].unique())

['Diesel' 'Petrol' 'LPG' 'CNG']
['Individual' 'Dealer' 'Trustmark Dealer']
['Manual' 'Automatic']
['First Owner' 'Second Owner' 'Third Owner' 'Fourth & Above Owner'
 'Test Drive Car']


In [87]:
df3[df3.duplicated()]

year  selling_price  km_driven    fuel seller_type transmission  \
291   2017         450000      35000  Petrol  Individual       Manual   
296   2012         330000      50000  Petrol  Individual       Manual   
370   2017        2625000       9000  Diesel      Dealer    Automatic   
371   2019        5150000      20000  Petrol      Dealer    Automatic   
372   2017        3200000      45000  Diesel      Dealer    Automatic   
...    ...            ...        ...     ...         ...          ...   
7984  2018        1265000      12000  Diesel  Individual       Manual   
7985  2019        1025000      32000  Diesel  Individual       Manual   
8114  2015         625000      50000  Diesel  Individual       Manual   
8123  2013         290000      25000  Diesel  Individual       Manual   
8124  2013         290000      25000  Diesel  Individual       Manual   

             owner  seats  Mileage  Engine_CC  Max_power  Torque     RPM  
291    First Owner    5.0    18.90       1197      82.00   114.0  4000.0  
296   Second Owner    5.0    18.60       1197      85.80   114.0  4000.0  
370    First Owner    5.0    13.60       1999     177.00   430.0  2125.0  
371    First Owner    5.0    22.37       2487     214.56   202.0  4400.0  
372    First Owner    5.0    19.33       1999     177.00   430.0  2125.0  
...            ...    ...      ...        ...        ...     ...     ...  
7984   First Owner    5.0    20.37       1461     108.45   240.0  1750.0  
7985   First Owner    5.0    28.09       1248      88.50   200.0  1750.0  
8114   First Owner    5.0    26.59       1248      74.00   190.0  2000.0  
8123   First Owner    5.0    23.57       1396      70.00   140.0  2400.0  
8124   First Owner    5.0    23.57       1396      70.00   140.0  2400.0  

[1218 rows x 13 columns]

In [88]:
df4 = df3.copy()

In [91]:
df4 = df4.drop_duplicates(keep='first')

In [93]:
df4.shape

(6906, 13)

In [94]:
X = df4.drop("selling_price", axis = 1)
y = df4["selling_price"]

In [95]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [99]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

categorical_cols = X_train.select_dtypes(include="object").columns
numerical_cols = X_train.select_dtypes(exclude = "object").columns

numeric_pipeline = Pipeline(
    steps = [
        ("scaler", StandardScaler())
    ]
)

categorical_pipeline = Pipeline(
    steps = [
        ("onehotencoder", OneHotEncoder(handle_unknown="ignore", drop = "first", sparse_output=False)),
        ("scaler", StandardScaler())
    ]
)

preprocessor = ColumnTransformer(
    [
        ("numeric_pipeline", numeric_pipeline, numerical_cols),
        ("categorical_pipeline", categorical_pipeline, categorical_cols)
    ],
    remainder = "passthrough"
)

In [100]:
X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)

In [103]:
X_train_processed[0]

array([-0.34638323,  0.74926534,  2.62890241, -1.86255564,  1.53758468,
        1.01680356, -0.10300134, -0.50688028,  0.91623896, -0.07752312,
       -0.8914437 ,  0.33887092, -0.0555606 ,  0.30742738, -0.15827517,
        1.56604598, -0.03009919, -0.29236703])

## Model Creation

In [105]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, AdaBoostRegressor, ExtraTreesRegressor
from xgboost import XGBRegressor
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [106]:
models = {
    "Linear Regression" : LinearRegression(),
    "Ridge" : Ridge(),
    "Lasso" : Lasso(),
    "SVR" : SVR(),
    "KNN" : KNeighborsRegressor(),
    "Decision Tree" : DecisionTreeRegressor(),
    "Random Forest" : RandomForestRegressor(),
    "Gradient Boosting" : GradientBoostingRegressor(),
    "AdaBoost" : AdaBoostRegressor(),
    "Extra Trees" : ExtraTreesRegressor(),
    "XGBoost" : XGBRegressor()
}
models

{'Linear Regression': LinearRegression(),
 'Ridge': Ridge(),
 'Lasso': Lasso(),
 'SVR': SVR(),
 'KNN': KNeighborsRegressor(),
 'Decision Tree': DecisionTreeRegressor(),
 'Random Forest': RandomForestRegressor(),
 'Gradient Boosting': GradientBoostingRegressor(),
 'AdaBoost': AdaBoostRegressor(),
 'Extra Trees': ExtraTreesRegressor(),
 'XGBoost': XGBRegressor(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_stra

In [107]:
for i in range(len(list(models))):
    model = list(models.values())[i]
    model.fit(X_train_processed, y_train)

    y_pred_train = model.predict(X_train_processed)
    y_pred_test = model.predict(X_test_processed)

    model_train_r2 = r2_score(y_train, y_pred_train)
    model_test_r2 = r2_score(y_test, y_pred_test)

    print(f"{list(models.keys())[i]} training accuracy: {model_train_r2 * 100}")
    print(f"{list(models.keys())[i]} testing accuracy: {model_test_r2 * 100}")
    print("="*35)
    print("\n")

Linear Regression training accuracy: 66.02091545758388
Linear Regression testing accuracy: 59.61151398154904


Ridge training accuracy: 66.020912601758
Ridge testing accuracy: 59.61422120014884


Lasso training accuracy: 66.02091544364956
Lasso testing accuracy: 59.611558146944745


SVR training accuracy: -5.138310438852733
SVR testing accuracy: -4.68055263062519


KNN training accuracy: 87.93334702720499
KNN testing accuracy: 77.76809959368973


Decision Tree training accuracy: 99.93090772653167
Decision Tree testing accuracy: 83.41593370415742


Random Forest training accuracy: 98.65594246492134
Random Forest testing accuracy: 89.73415024458556


Gradient Boosting training accuracy: 94.35212988409144
Gradient Boosting testing accuracy: 87.63675114111172


AdaBoost training accuracy: 41.18325002129419
AdaBoost testing accuracy: 8.304500980579322


Extra Trees training accuracy: 99.93071843338784
Extra Trees testing accuracy: 88.98425434272885


XGBoost training accuracy: 98.7276852130

## Best performing model is random forest with a accuracy of 89.73%.